### PR TITLE
Global filter box

### PR DESCRIPTION
**Pull request by chowington**
Fixes https://github.com/VEuPathDB/web-eda/issues/175

### CODE REVIEW

**Review by dmfalke — COMMENTED** [review]
I think this looks really great! I have three suggestions:
1. Make the dialog bound to the viewport. Currently, it's possible to move the dialog such that the title bar is not accessible and the only way to close it is by clicking the "See all filters" button to hide it.
2. Remember the dimensions and position of the dialog box so that it is in the same place the next time it is opened.
3. Use the `contained` variable of the "See all filters" button so that it looks more button-like, and change the text to "Active filters" so that it is the same as the title bar of the dialog. I think this will help user better understand the relationship of the button and the dialog.

What do you think?

**Review by bobular — COMMENTED** [review]
I also think it looks great.

Question, not opinion, does the dialogue need to be relative to the browser viewport, or scroll up and down with the rest of the page?

Possibly a Phase 2 thing, but I'd really like an undo button, or two actions on a filter chip:

1. reversibly disable
2. delete

File: src/lib/core/components/GlobalFiltersDialog.tsx
Comment (jtlong3rd) [inline src/lib/core/components/GlobalFiltersDialog.tsx:36]: Just as an FYI, in cases like these, it's more idiomatic to use [Array.prototype.reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce) instead of a `for` loop.

("cases like these" = "cases where you're iterating through each `element` of an `array` and updating/mutating some sort of `value` based on each `element`" In this case, you're iterating through each `entity` and updating `filters`, `matchingFilters`, and `filterChipLists` based on each of those `entity`s.)

I don't think it's necessary to hold up the review/merge over this, so after the merge, I'll leave a comment here showing how this `for` loop could have been refactored to use `reduce`. That'll make for a good reference for how you could use `reduce` in future work!

**Review by jtlong3rd — COMMENTED** [review]
The new functionality works great! No suggestions for different UX/styling. I have a few minor suggestions for the implementation, one of which doesn't need to be addressed right now.

File: src/lib/core/components/GlobalFiltersDialog.tsx
Comment (bobular) [reply src/lib/core/components/GlobalFiltersDialog.tsx:36]: https://www.youtube.com/watch?v=qaGjS7-qWzg ;-)

**Review by jtlong3rd — APPROVED** [review]

**Comment by chowington** [comment]
> 
> 
> I think this looks really great! I have three suggestions:
> 
>     1. Make the dialog bound to the viewport. Currently, it's possible to move the dialog such that the title bar is not accessible and the only way to close it is by clicking the "See all filters" button to hide it.
> 
>     2. Remember the dimensions and position of the dialog box so that it is in the same place the next time it is opened.
> 
>     3. Use the `contained` variable of the "See all filters" button so that it looks more button-like, and change the text to "Active filters" so that it is the same as the title bar of the dialog. I think this will help user better understand the relationship of the button and the dialog.
> 
> 
> What do you think?

Thanks for the feedback Dave!

3. Done in accordance with the discussion at our last EDA dev meeting 👍 

1 & 2. I implemented these but I'm not super happy with them. Side effects tied to #1 cause the odd behavior of resizing the dialog to expand all sides simultaneously (that is, the box center stays in place) rather than the usual expansion of the resizing corner. This also means that you can accidentally lose the header offscreen when resizing, losing the ability to drag unless you resize again. However because you can also resize the box such that the resizing handle itself is offscreen (don't see a simple solution to that), it's just a bit easier to get in a predicament. #2 also creates the difficulty that if you do get in a sticky situation with the sizing/position of the box, you can't simply close and reopen it to reset everything back to normal. I'll let y'all have a think and play on it to form your own opinion.

**Comment by chowington** [comment]
> 
> 
> I also think it looks great.
> 
> Question, not opinion, does the dialogue need to be relative to the browser viewport, or scroll up and down with the rest of the page?
> 
> Possibly a Phase 2 thing, but I'd really like an undo button, or two actions on a filter chip:
> 
>     1. reversibly disable
> 
>     2. delete

Thanks Bob! I think the disable chip option is a good idea for phase 2 :)

File: src/lib/core/components/FilterChipList.tsx
Comment (chowington) [reply src/lib/core/components/FilterChipList.tsx]: > 
> 
> The new functionality works great! No suggestions for different UX/styling. I have a few minor suggestions for the implementation, one of which doesn't need to be addressed right now.

Thanks Jamie!

Comment (chowington) [reply src/lib/core/components/FilterChipList.tsx]: > 
> 
> To find the entity and variable, I would use the existing `findEntityAndVariable` utility in `utils/study-metadata`:
> 
> ```ts
> const { entity, variable } = findEntityAndVariable(props.entities, filter) ?? {};
> ```

Good idea! Done.

File: src/lib/workspace/AnalysisPanel.tsx
Comment (chowington) [reply src/lib/workspace/AnalysisPanel.tsx]: Good idea as well! Done.

File: src/lib/core/components/GlobalFiltersDialog.tsx
Comment (chowington) [reply src/lib/core/components/GlobalFiltersDialog.tsx]: These often slip away from me, so thanks for the reminder! Done

File: src/lib/core/components/GlobalFiltersDialog.tsx
Comment (chowington) [reply src/lib/core/components/GlobalFiltersDialog.tsx:39]: Thanks for the suggestion, Jamie! I'm familiar with `reduce`. Bob's counterpoint (implied, in video form) is worth considering. I think it's an interesting question as to whether `reduce` in this instance would improve the code. I'm open!

**Comment by dmfalke** [comment]
> 1 & 2. I implemented these but I'm not super happy with them. Side effects tied to No.1 cause the odd behavior of resizing the dialog to expand all sides simultaneously (that is, the box center stays in place) rather than the usual expansion of the resizing corner. This also means that you can accidentally lose the header offscreen when resizing, losing the ability to drag unless you resize again. However because you can also resize the box such that the resizing handle itself is offscreen (don't see a simple solution to that), it's just a bit easier to get in a predicament. No.2 also creates the difficulty that if you do get in a sticky situation with the sizing/position of the box, you can't simply close and reopen it to reset everything back to normal. I'll let y'all have a think and play on it to form your own opinion.

Hi @chowington, I pushed a commit to address the resize issue. Let me know what you think about it when you're back from vacation.

**Comment by chowington** [comment]
> 
> 
> > 1 & 2. I implemented these but I'm not super happy with them. Side effects tied to No.1 cause the odd behavior of resizing the dialog to expand all sides simultaneously (that is, the box center stays in place) rather than the usual expansion of the resizing corner. This also means that you can accidentally lose the header offscreen when resizing, losing the ability to drag unless you resize again. However because you can also resize the box such that the resizing handle itself is offscreen (don't see a simple solution to that), it's just a bit easier to get in a predicament. No.2 also creates the difficulty that if you do get in a sticky situation with the sizing/position of the box, you can't simply close and reopen it to reset everything back to normal. I'll let y'all have a think and play on it to form your own opinion.
> 
> Hi @chowington, I pushed a commit to address the resize issue. Let me know what you think about it when you're back from vacation.

Thanks Dave, it looks good!

**Pull request merged**
